build/xtask/src/flash.rs

```diff
@@ -11,59 +12,233 @@ use anyhow::Context;
 
 use crate::Config;
 
-pub fn run(verbose: bool, cfg: &Path) -> anyhow::Result<()> {
-    ctrlc::set_handler(|| {}).expect("Error setting Ctrl-C handler");
+//
+// We allow for enough information to be put in the archive for the image to
+// be flashed based only on the archive (e.g., by Humility).  Because flashing
+// is itself a bit of a mess (requiring different programs for different
+// targets), this is a bit gritty (e.g., any required external configuration
+// files must themselves put in the archive).  If these structures need to
+// change, be sure to make corresponding changes to Humility.
+//
+#[derive(Debug, Serialize)]
+pub enum FlashProgram {
+    PyOcd(Vec<FlashArgument>),
+    OpenOcd(FlashProgramConfig),
+}
 
-    let toml = Config::from_file(&cfg)?;
+//
+// Enum describing flash programs configuration (e.g., "openocd.cfg" for
+// OpenOCD), either as a path in the file system or with the entire contents.
+//
+#[derive(Debug, Serialize)]
+pub enum FlashProgramConfig {
+    Path(Vec<String>),
+    Payload(String),
+}
 
-    let mut out = PathBuf::from("target");
-    out.push(toml.name);
-    out.push("dist");
+//
+// An enum describing a single command-line argument to the flash program.
+//
+#[derive(Debug, Serialize)]
+pub enum FlashArgument {
+    // A direct string
+    Direct(String),
 
-    let (mut flash, reset) = match toml.board.as_str() {
-        "lpcxpresso55s69" => {
-            let mut flash = Command::new("pyocd");
-            flash
-                .arg("flash")
-                .arg("-t")
-                .arg("lpc55s69")
-                .arg("--format")
-                .arg("hex")
-                .arg(out.join("final.ihex"));
+    // The filesystem path of the binary flash payload itself
+    Payload,
 
-            let mut reset = Command::new("pyocd");
-            reset.arg("reset").arg("-t").arg("lpc55s69");
+    // A single argument consisting of a prefix, the path of the flash
+    // payload, and a suffix
```

Comment:
There are only two things here, not three (and it looks like we hardcode `final.srec` as the payload name?)

Reply:
Will clarify the comment -- it's up to the user of this thing to inject the path to the payload in the middle.